### ISSUE TITLE
SEC-1759: AbstractPreAuthenticatedProcessingFilter should be more subclass-friendly

### ISSUE THREAD

**Issue by spring-issuemaster** [spring-projects/spring-security#1977]
[Vasily Ivanov](https://jira.spring.io/secure/ViewProfile.jspa?name=bas) (Migrated from [SEC-1759](https://jira.spring.io/browse/SEC-1759?redirect=false)) said:

I had a requirement to handle different mutually exclusive pre-authentication mechanisms in the same app.

One puts pre-authenticated principal into request header (so RequestHeaderAuthenticationFilter worked out of the box for me, it simply returns request header value as String).
Another one uses secret request query parameter as authentication principal (security through obscurity) so I just extended AbstractPreAuthenticatedProcessingFilter and implemented getPreAuthenticatedPrincipal() method that simply returns query parameter value as String.

Then I chained this two filters in my security config and it all worked good until I realised that I need to load authenticated user details differently for different  pre-authentication scenarios... so I couldn't work with String principals anymore as String is ambiguous.

In my filters I had to wrap String principal into new class PreAuthenticatedPrincipal and add source of pre-authentication to it, so AuthenticationUserDetailsService can distinguish between pre-auth providers nicely and act accordingly. So far so good.

While doing all that I came across problem of identifying principal changes. AbstractPreAuthenticatedProcessingFilter's requiresAuthentication() method does currentUser.getName().equals(principal) which _always_ returns false on non-String principals. I wanted to extend it, but I couldn't as both requiresAuthentication and doAuthenticate are private and not extendable. Also on principal change I want to do a bit more then just invalidating current session which is not possible with current AbstractPreAuthenticatedProcessingFilter implementation. IMO it needs to be more like LogoutFilter in the way it supports a number of logout handlers, smart defaults, etc.

So basically I need AbstractPreAuthenticatedProcessingFilter to define requiresAuthentication as protected (at least), so I can implement all that myself.

**Comment by eleftherias** [comment]
Fixed in 62c7de0.
